### DISCUSSION TITLE
Option for sending or receiving locked rotor signal

### DISCUSSION THREAD

**Discussion by mjpcomp** [tjko/fanpico#86] [General]
Some fans, instead of a tach or speed control, send a locked rotor signal on that 3rd wire. Is there a way to change the signal that connect back to the motherboard to be a locked rotor signal?

Example situation:
Many network/server devices rely on that signal, however, those fans are harder to come by than a PWM or a Tach fan. Would be nice if we can use fanpico to connect fans that provide a speed/PWM control signal, and on the connection back to the motherboard/network device, send a high (or low, configurable) signal based on a configurable tach/speed signal.

Let's say that we have fans attached to the fanpico that operate up to (Example) 6000 RPM... If the speed drops below 500 RPM (configurable), send a high (or low) signal back to the motherboard/device so it knows there is a fan error and can do whatever it needs to take corrective actions (alerts/shutdown/whatever that device would do on a fan failure).

**Comment by mjpcomp** [comment]
Conversely, if we have locked rotor signal fans, and need to send a speed down to the attached motherboard/device, would that be possible? In this situation, we may have the fan data sheet and (e.g. runs at 6K RPM) so we send a false 6000 RPM signal down to the end device. If the fan locks up and sends us a high/low signal on the 3rd wire, report 0 RPM to the connected device.

**Comment by tjko** [comment]
It should be relatively easy to add support for both cases:
- Interpreting locked rotor signal received from fan (perhaps mapping it to user specified RPM values 0 and x RPM, etc...)
- Configuring MBFAN ports to either send normal ~~PWM~~ tachometer signal or locked rotor signal (back to motherboard)

This should allow using regular PWM fans with device that expects to see rotor locked signal (and vice versa).  

And this should still work with the existing functionality (CONF:MBFANx:SOURCE) to send "MIN" tachometer signal from multiple fans back to one motherboard fan connector...

**Reply by tjko** [reply]
It probably would be possible to "auto-detect" if fan is sending locked rotor signal or not, but maybe better just add commands to specify the type of "tachometer" signal Fan is sending (and what type output to send on tachometer pin on the _MBFAN_ outputs).

I was thinking that following commands should allow supporting all kinds of configurations?


**CONFigure:FANx:RPMMode <mode>**

where _mode_ would be:

TACH = fan is sending normal tachometer pulses
LR,0,6000 = fan is sending locked rotor signal that is mapped to 0 rpm (HIGH) and 6000 rpm (LOW)


**CONFigure:MBFANx:RPMMode <mode>**

where _mode_ is:

TACH = fanpico generates tachometer pulses (to simulate regular fan tachometer output)
LR,500,HIGH = fanpico sends locked rotor signal (HIGH) if RPM < 500

**Reply by mjpcomp** [reply]
> It probably would be possible to "auto-detect" if fan is sending locked rotor signal or not, but maybe better just add commands to specify the type of "tachometer" signal Fan is sending (and what type output to send on tachometer pin on the _MBFAN_ outputs).
> 
> I was thinking that following commands should allow supporting all kinds of configurations?
> 
> **CONFigure:FANx:RPMMode **
> 
> where _mode_ would be:
> 
> TACH = fan is sending normal tachometer pulses LR,0,6000 = fan is sending locked rotor signal that is mapped to 0 rpm (HIGH) and 6000 rpm (LOW)
> 
> **CONFigure:MBFANx:RPMMode **
> 
> where _mode_ is:
> 
> TACH = fanpico generates tachometer pulses (to simulate regular fan tachometer output) LR,500,HIGH = fanpico sends locked rotor signal (HIGH) if RPM < 500

That would be absolutely amazing if it could be implemented like that. I'm guessing a whole subreddit or two would want to grab a board or two for their switches to quiet them down with an expanded range of fans that don't need to be LR fans. I know I have bunch of switches that I'd be getting some boards for ASAP.

And thanks for that quick reply 👍

**Comment by tjko** [comment]
I added initial support for Locked Rotor (Alarm) fans. Code is in its own branch (until I have time to test/validate that it works as expected): https://github.com/tjko/fanpico/tree/lockedrotor

If someone want's to help and test/validate locked rotor fan support, pre-built test firmware can be found here: https://github.com/tjko/fanpico/actions/runs/9690054746

**Reply by mjpcomp** [reply]
That's excellent - placing an order for one of the kits right now on eBay, so I'll definitely try it out once I get and assemble it.

**Reply by mjpcomp** [reply]
Got the tracking number, so, now just waiting for them to come in, and then can assemble and try it out

**Comment by tjko** [comment]
Initial support for locked rotor (alarm) signal fans has now been merged (https://github.com/tjko/fanpico/commit/1cf7e62aaad28090545bd451102735728e5802b0)

Pre-built test firmware images can be found here: https://github.com/tjko/fanpico/actions/runs/9707277191

**Comment by tjko** [comment]
I found a Locked Rotor Alarm fan (NMB-MAT 3110RL-04W-S59), and it seems to work fine with FanPico, if one sets fan RPMMODE to ```LRA,1000,0``` (locked rotor signal is low when fan operates and if fan stops signal goes high).

Here is signal change captured when rotor is stopped:
![LockedRotorSignalChangeWhenRotorStops](https://github.com/tjko/fanpico/assets/1075726/3c64baa6-ee47-47d4-91a3-a2de14650eb7)


Is there 'standard' on the locked rotor signal (polarity)?  Current defaults in Fanpico are to assume 0 RPM if signal is low, and 1000 RPM when signal is high. Should the default perhaps be reversed?

**Reply by mjpcomp** [reply]
Thanks for going the extra mile to test it - I received the kits, but, have a sick daughter so haven't been able to solder the kits together.

In terms of high, low, there isn't a standard it seems. Though many of the fans I've looked at report high (at least on the data sheets) when the rotor is locked, so, 0 RPM = High. I'm guessing that's more of the norm.
